ts41/get.d.ts

```diff
@@ -0,0 +1,123 @@
+import {Split} from './utilities';
+import {StringDigit} from '../source/utilities';
+
+/**
+Like @see Get but receives an array of strings as a path parameter.
+*/
+type GetWithPath<ObjectType, Keys extends readonly string[]> =
+	Keys extends []
+	? ObjectType
+	: Keys extends [infer Head, ...infer Tail]
+	? GetWithPath<PropertyOf<ObjectType, Extract<Head, string>>, Extract<Tail, string[]>>
+	: never;
+
+/**
+Splits a dot-prop style path into a tuple comprised of the properties in the path. Handles square-bracket notation.
+
+@example
+```
+ToPath<'foo.bar.baz'>
+//=> ['foo', 'bar', 'baz']
+
+ToPath<'foo[0].bar.baz'>
+//=> ['foo', '0', 'bar', 'baz']
+```
+*/
+type ToPath<S extends string> = Split<FixPathSquareBrackets<S>, '.'>;
+
+/**
+Replaces square-bracketed dot notation with dots, for example, `foo[0].bar` -> `foo.0.bar`.
+*/
+type FixPathSquareBrackets<Path extends string> =
+	Path extends `${infer Head}[${infer Middle}]${infer Tail}`
+	? `${Head}.${Middle}${FixPathSquareBrackets<Tail>}`
+	: Path;
+
+/**
+Returns true if `S` is made up out of `C` repeated 0 or more times.
+
+@example
+```
+ConsistsOnlyOf<'aaa', 'a'> //=> true
+ConsistsOnlyOf<'aBa', 'a'> //=> false
+ConsistsOnlyOf<'', 'a'> //=> true
+```
+*/
+type ConsistsOnlyOf<LongString extends string, Substring extends string> =
+	LongString extends ''
+	? true
+	: LongString extends `${Substring}${infer Tail}`
+	? ConsistsOnlyOf<Tail, Substring>
+  : false;
+
+/**
+Convert a type which may have number keys to one with string keys, making it possible to index using strings retrieved from template types.
+
+@example
+```
+type WithNumbers = {foo: string; 0: boolean};
+type WithStrings = WithStringKeys<WithNumbers>;
+
+type WithNumbersKeys = keyof WithNumbers;
+//=> 'foo' | 0
+type WithStringsKeys = keyof WithStrings;
+//=> 'foo' | '0'
+```
+*/
+type WithStringKeys<ObjectType extends Record<string | number, any>> = {
+	[Key in `${Extract<keyof ObjectType, string | number>}`]: ObjectType[Key]
+};
+
+/**
+Get a property of an object or array. Works when indexing arrays using number-literal-strings, for example, `PropertyOf<number[], '0'> = number`, and when indexing objects with number keys.
+
+Returns `unknown` if `Key` is not a property of `ObjectType`.
+*/
+type PropertyOf<ObjectType, Key extends string> =
+	ObjectType extends null | undefined
+	? undefined
+	: Key extends keyof ObjectType
+	? ObjectType[Key]
+	: ObjectType extends ArrayLike<infer Item>
+	? ConsistsOnlyOf<Key, StringDigit> extends true
+	? Item
+	: unknown
```

Comment:
We needs some parenthesis here to make things readable.

Only when the rows alternate between `?` and `:` every other row can parenthesis be skipped, as they are then essentially the `else if`:s of ternary operators. But this isn't an `else if`, this is a nested `if`, so we need parenthesis to highlight that and separate it from the larger `else if`-style ternary:

```suggestion
	? (
		ConsistsOnlyOf<Key, StringDigit> extends true
		? Item
		: unknown
	)
```

Also: `unknown` here should be `never`?

Reply:
Will improve the formatting. @sindresorhus do you generally use prettier or anything similar? I'd love to have the decision taken out of my hands!

Re `unknown` vs `never` - similar to the other comment - `never` would imply an error being thrown when the key isn't found. All of this assumes that we're trying to modal a deep-get function as accurately as possible. There could definitely be type-land use-cases for `never`s - maybe `PathKey` and `PathValue` from the playground could co-exist with this type?

Reply:
> Will improve the formatting. @sindresorhus do you generally use prettier or anything similar? I'd love to have the decision taken out of my hands!

I don't. Prettier is a good idea in theory, but in practice, it makes too many weird decisions, often making code less readable.